dm/worker/log.go

```diff
@@ -0,0 +1,498 @@
+	Location int64
+}
+
+// MarshalBinary never return not nil err now
+func (p *Pointer) MarshalBinary() []byte {
+	data := make([]byte, 8)
+	binary.LittleEndian.PutUint64(data, uint64(p.Location))
+
+	return data
+}
+
+// UnmarshalBinary implement encoding.BinaryMarshal
+func (p *Pointer) UnmarshalBinary(data []byte) error {
+	if len(data) < 8 {
+		return errors.Errorf("no enough data as pointer %v", data)
+	}
+
+	p.Location = int64(binary.LittleEndian.Uint64(data))
+	return nil
+}
+
+// LoadHandledPointer loads handled pointer value from kv DB
+func LoadHandledPointer(db *leveldb.DB) (Pointer, error) {
+	var p Pointer
+	if db == nil {
+		return p, errors.Trace(ErrInValidHandler)
+	}
+
+	value, err := db.Get(HandledPointerKey, nil)
+	if err != nil {
+		// return zero value when not found
+		if err == leveldb.ErrNotFound {
+			return p, nil
+		}
+
+		return p, errors.Trace(err)
+	}
+
+	err = p.UnmarshalBinary(value)
+	if err != nil {
+		return p, errors.Trace(err)
+	}
+
+	return p, nil
+}
+
+var (
+	defaultGCForwardLog int64 = 10000
+
+	// TaskLogPrefix is  prefix of task log key
+	TaskLogPrefix = []byte("!DM!TaskLog")
+)
+
+// DecodeTaskLogKey decodes task log key and returns its log ID
+func DecodeTaskLogKey(key []byte) (int64, error) {
+	if len(key) != len(TaskLogPrefix)+8 {
+		return 0, errors.Errorf("no enough data as task log key %v", key)
+	}
+
+	return int64(binary.LittleEndian.Uint64(key[len(TaskLogPrefix):])), nil
+}
+
+// EncodeTaskLogKey encodes log ID into a task log key
+func EncodeTaskLogKey(id int64) []byte {
+	key := make([]byte, 8+len(TaskLogPrefix))
+	copy(key[:len(TaskLogPrefix)], TaskLogPrefix)
+	binary.LittleEndian.PutUint64(key[len(TaskLogPrefix):], uint64(id))
+
+	return key
+}
+
+// Logger manage task operation logs
+type Logger struct {
+	endPointer     Pointer
+	handledPointer Pointer
+}
+
+// Initial initials Logger
+func (logger *Logger) Initial(db *leveldb.DB) ([]*pb.TaskLog, error) {
+	if db == nil {
+		return nil, errors.Trace(ErrInValidHandler)
+	}
+
+	handledPointer, err := LoadHandledPointer(db)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var (
+		endPointer = Pointer{
+			Location: handledPointer.Location + 1,
+		}
+		logs = make([]*pb.TaskLog, 0, 4)
+	)
+	iter := db.NewIterator(util.BytesPrefix(TaskLogPrefix), nil)
+	startLocation := handledPointer.Location + 1
+	for ok := iter.Seek(EncodeTaskLogKey(startLocation)); ok; ok = iter.Next() {
+		logBytes := iter.Value()
+		log := &pb.TaskLog{}
+		err = log.Unmarshal(logBytes)
+		if err != nil {
+			err = errors.Annotatef(err, "unmarshal task log %s", logBytes)
+			break
+		}
+
+		if log.Id >= endPointer.Location {
+			// move to next location
+			endPointer.Location = log.Id + 1
+			logs = append(logs, log)
+		} else {
+			panic("out of sorted from level db")
+		}
+	}
+	iter.Release()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	err = iter.Error()
+	if err != nil {
+		return nil, errors.Annotatef(err, "fetch logs from meta")
```

Comment:
```suggestion
		return nil, errors.Annotate(err, "fetch logs from meta")
```

Reply:
It's also ok

Reply:
yeh, only one unnecessary call to `fmt.Sprintf`.